packages/main/src/Calendar.ts

```diff
@@ -285,41 +294,104 @@ class Calendar extends CalendarPart {
 	@property({ type: CalendarLegendItemType, defaultValue: CalendarLegendItemType.None })
 	_selectedItemType!: `${CalendarLegendItemType}`;
 
+	constructor() {
+		super();
+
+		this._valueIsProcessed = false;
+	}
+
 	/**
 	 * @private
 	 */
 	get _selectedDatesTimestamps(): Array<number> {
-		return this.dates.map(date => {
-			const value = date.value;
-			const validValue = value && !!this.getFormat().parse(value);
-			return validValue ? this._getTimeStampFromString(value)! / 1000 : undefined;
-		}).filter((date): date is number => !!date);
-	}
+		let selectedDates: Array<number> = [];
 
-	constructor() {
-		super();
+		switch (this.selectionMode) {
+		case CalendarSelectionMode.Range: {
+			const range = this.dates.filter(date => date.hasAttribute("ui5-date-range"))[0];
+			const startDate = range && range.startValue && this.getFormat().parse(range.startValue, true) as Date;
+			const endDate = range && range.endValue && this.getFormat().parse(range.endValue, true) as Date;
 
-		this._valueIsProcessed = false;
+			if (startDate) {
+				selectedDates.push(startDate.getTime() / 1000);
+			}
+
+			if (endDate) {
+				selectedDates.push(endDate.getTime() / 1000);
+			}
+			break;
+		}
+		case CalendarSelectionMode.Multiple:
+		default: {
+			selectedDates = this.dates
+				.filter(dateElement => {
+					return dateElement.hasAttribute("ui5-date")
+						&& dateElement.value
+						&& this._isValidCalendarDate(dateElement.value)
+						&& this._getTimeStampFromString(dateElement.value);
+				})
+				.map(dateElement => Number(this._getTimeStampFromString(dateElement.value!)) / 1000);
+			break;
+		}
+		}
+
+		return selectedDates;
 	}
 
 	/**
 	 * @private
 	 */
 	_setSelectedDates(selectedDates: Array<number>) {
-		const selectedValues = selectedDates.map(timestamp => this.getFormat().format(UI5Date.getInstance(timestamp * 1000), true)); // Format as UTC
-		const valuesInDOM = [...this.dates].map(dateElement => dateElement.value);
-
-		// Remove all elements for dates that are no longer selected
-		this.dates.filter(dateElement => !selectedValues.includes(dateElement.value)).forEach(dateElement => {
-			this.removeChild(dateElement);
-		});
-
-		// Create tags for the selected dates that don't already exist in DOM
-		selectedValues.filter(value => !valuesInDOM.includes(value)).forEach(value => {
-			const dateElement = document.createElement(CalendarDate.getMetadata().getTag()) as CalendarDate;
-			dateElement.value = value;
-			this.appendChild(dateElement);
-		});
+		const selectedUTCDates = selectedDates.map(timestamp => this.getFormat().format(UI5Date.getInstance(timestamp * 1000), true));
+
+		switch (this.selectionMode) {
+		case CalendarSelectionMode.Range: {
+			// Create tags for the selected dates that don't already exist in DOM
+			if (selectedUTCDates.length) {
+				let dateRange = this.dates.find(dateElement => dateElement.startValue === selectedUTCDates[0]) as CalendarDateRange;
+				if (!dateRange) {
```

Comment:
Find from previous rows might return undefined so casting to CalendarDateRange type will exclude that and casting it to the class again when it wasn't found seems conceptually wrong for me. Maybe the check is not correct?

You can check also TypeScript type guards which could possibly remove many of the casts.

Reply:
I changed it to check for ui5-date-ranges specifically now.